### PR TITLE
fix: filter out Symbol.hasInstance

### LINKED ISSUE #1671: [0.18.1] TypeError: Symbol.hasInstance is read-only
### Describe the bug

My tests that use `vi.mock` (in this case, mocking out the `ora` module) started failing as of vitest v0.18.1 with the following error:

```
TypeError: Cannot assign to read only property 'Symbol(Symbol.hasInstance)' of function 'function(...o) {
    if (e.called = !0, e.callCount++, e.calls.push(o), e.next) {
      let [s, l] = e.next;
  ...<omitted>... }'
```

In the reproduction (below), it presents itself as simply: `TypeError: Symbol.hasInstance is read-only`

I'm assuming this is related to PR #1648 but I'm not sure if I need to update my usage to comply with the intent of the PR or what.

### Reproduction

https://stackblitz.com/edit/vitest-dev-vitest-ht4a9x?file=test/mock.test.ts&view=editor

### System Info

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (16) x64 Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
    Memory: 17.74 GB / 31.92 GB
  Binaries:
    Node: 16.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.18 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.12.2 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: 103.0.5060.114
    Edge: Spartan (44.19041.1266.0), Chromium (103.0.1264.62)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    vitest: 0.18.1 => 0.18.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by nieyuyao**
fix #1671

### CODE REVIEW

**Review by QGerrit — APPROVED** [review]

**Review by QGerrit — COMMENTED** [review]
This unfortunately doesn't seem to quite be enough to fix #1671.... at least fully. After manually applying this patch locally, my tests started failing with `TypeError: Cannot read properties of undefined (reading 'dataEmitted')`. `jsdom` test environment in a very basic vue3 app.... oddly, `dataEmitted` doesn't seem to appear anywhere in my node_modules tree.

**Comment by nieyuyao** [comment]
> This unfortunately doesn't seem to quite be enough to fix #1671.... at least fully. After manually applying this patch locally, my tests started failing with `TypeError: Cannot read properties of undefined (reading 'dataEmitted')`. `jsdom` test environment in a very basic vue3 app.... oddly, `dataEmitted` doesn't seem to appear anywhere in my node_modules tree.

emmm.... I tried to find the reason and converted the PR into draft

**Comment by nieyuyao** [comment]
This `TypeError: Cannot read properties of undefined (reading 'dataEmitted')` comes from [here](https://github.com/nodejs/node/blob/948712bc4c4b24edc238222ba6a7e585f51d01ef/lib/internal/streams/readable.js#L1172). Try to read the readableDidRead property and assign to the container when mocking. However, the Readable has not been initialized yet, so an error is thrown.

**Comment by QGerrit** [comment]
Wow, good find! I forgot to mention I was doing `vi.mock('axios')`, that break location makes more sense now.

File: packages/vitest/src/runtime/mocker.ts [hunk 1]
```diff
@@ -32,8 +32,7 @@ type Key = string | symbol
 
 function isSpecialProp(prop: Key, parentType: string) {
   return parentType.includes('Function')
-      && typeof prop === 'string'
-      && ['arguments', 'callee', 'caller', 'length', 'name'].includes(prop)
+      && ['arguments', 'callee', 'caller', 'length', 'name', Symbol.hasInstance].includes(prop)
```
Comment (simon-abbott) [inline packages/vitest/src/runtime/mocker.ts:35]: I don't like this solution. There are valid cases where we _do_ want to mock `Symbol.hasInstance`, and, as you've noticed, this doesn't actually fix the root cause.

**Review by simon-abbott — CHANGES_REQUESTED** [review]
I don't like this solution. There are valid cases where we _do_ want to mock `Symbol.hasInstance`, and, as you've noticed, this doesn't actually fix the root cause.

I'm working on a fix for the underlying issue which should be up shortly.

**Comment by simon-abbott** [comment]
I've solved #1671 in a more general way over in #1786.

**Pull request closed without merging**